File: core/src/ru/myitschool/dcrawler/effects/CellEffect.java
package ru.myitschool.dcrawler.effects;

import ru.myitschool.dcrawler.entities.Entity;

/**
 * Created by Voyager on 01.12.2017.
 */
public abstract class CellEffect extends Effect {

    FloorEffect floorEffect;

    public CellEffect(FloorEffect floorEffect) {
        super(null);
        this.floorEffect = floorEffect;
    }

    public FloorEffect getFloorEffect() {
        return floorEffect;
    }

    public void setFloorEffect(FloorEffect floorEffect) {
        this.floorEffect = floorEffect;
    }

    public void stepToAction(int x1, int y1, int x2, int y2, Entity entity){
        if (!hasCell(x1, y1) && hasCell(x2, y2)){
            onStepTo(entity);
            System.out.println("TO ZONE!!!");
        } else if (hasCell(x1, y1) && hasCell(x2, y2)){
            onMovingInZone(entity);
            System.out.println("IN ZONE!!!");
        }
    }

    protected void onStepTo(Entity entity){

    }

    protected void onMovingInZone(Entity entity){

    }

    public void stepFromAction(int x1, int y1, int x2, int y2, Entity entity){
        if (!hasCell(x2, y2) && hasCell(x1, y1)){
            onStepFrom(entity);
            System.out.println("FROM ZONE!!!");
        }
    }

    protected void onStepFrom(Entity entity){

    }

    private boolean hasCell(int x, int y) {
        return floorEffect.hasCell(x, y);
    }

    @Override
    public Object clone() {
        try {
            CellEffect effect = (CellEffect) super.clone();
            effect.setFloorEffect(getFloorEffect());
            return effect;
        } catch (Exception e){
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        CellEffect effect = (CellEffect) o;

        return floorEffect != null ? floorEffect.equals(effect.floorEffect) : effect.floorEffect == null;
    }
}

File: core/src/com/wolg_vlad/dcrawler/effects/CellEffect.kt
package com.wolg_vlad.dcrawler.effects

import com.wolg_vlad.dcrawler.entities.Entity

/**
 * Created by Voyager on 01.12.2017.
 */
abstract class CellEffect(var floorEffect: FloorEffect?) : Effect(null) {

    fun stepToAction(x1: Int, y1: Int, x2: Int, y2: Int, entity: Entity) {
        if (!hasCell(x1, y1) && hasCell(x2, y2)) {
            onStepTo(entity)
            println("TO ZONE!!!")
        } else if (hasCell(x1, y1) && hasCell(x2, y2)) {
            onMovingInZone(entity)
            println("IN ZONE!!!")
        }
    }

    protected open fun onStepTo(entity: Entity) {}
    protected fun onMovingInZone(entity: Entity) {}
    fun stepFromAction(x1: Int, y1: Int, x2: Int, y2: Int, entity: Entity) {
        if (!hasCell(x2, y2) && hasCell(x1, y1)) {
            onStepFrom(entity)
            println("FROM ZONE!!!")
        }
    }

    protected open fun onStepFrom(entity: Entity?) {}
    private fun hasCell(x: Int, y: Int): Boolean {
        return floorEffect!!.hasCell(x, y)
    }

    override fun clone(): Any {
        val effect = super.clone() as CellEffect
        effect.floorEffect = floorEffect
        return effect
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val effect = o as CellEffect
        return if (floorEffect != null) floorEffect == effect.floorEffect else effect.floorEffect == null
    }

    override fun hashCode(): Int {
        return floorEffect?.hashCode() ?: 0
    }

}